File: AndroidApp/app/src/main/java/com/splotus/smarthrms/WFHLocationService.kt
package com.splotus.smarthrms

import android.app.*
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.*
import android.util.Log
import androidx.core.app.NotificationCompat
import com.google.android.gms.location.*
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Foreground Service that runs background GPS tracking even when:
 * - App is in background
 * - Screen is locked / off
 * - Phone is on charge and idle
 *
 * Sends GPS pings to /api/wfh/locationPing every 20 seconds.
 * Shows a persistent notification (Android OS requirement for background services).
 */
class WFHLocationService : Service() {

    companion object {
        const val TAG = "WFHLocationService"

        // Intent actions
        const val ACTION_START = "com.splotus.smarthrms.START_TRACKING"
        const val ACTION_STOP  = "com.splotus.smarthrms.STOP_TRACKING"

        // Intent extras
        const val EXTRA_EMP_ID     = "emp_id"
        const val EXTRA_SERVER_URL = "server_url"

        const val NOTIFICATION_ID = 1001
    }

    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var locationCallback: LocationCallback
    private var locationCallbackRegistered = false

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .retryOnConnectionFailure(true)
        .build()

    // State
    private var empId: Int = 0
    private var serverUrl: String = ""
    private var lastLat: Double = 0.0
    private var lastLon: Double = 0.0
    private var pingCount: Int = 0
    private var lastPingSuccessful: Boolean = false

    // Wake lock to prevent CPU sleep between pings
    private var wakeLock: PowerManager.WakeLock? = null

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    override fun onCreate() {
        super.onCreate()
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        Log.i(TAG, "WFH Location Service created")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.i(TAG, "onStartCommand: action=${intent?.action}")

        when (intent?.action) {
            ACTION_START -> {
                empId     = intent.getIntExtra(EXTRA_EMP_ID, 0)
                serverUrl = intent.getStringExtra(EXTRA_SERVER_URL) ?: BuildConfig.SERVER_URL

                if (empId <= 0) {
                    Log.e(TAG, "Invalid empId. Stopping service.")
                    stopSelf()
                    return START_NOT_STICKY
                }

                acquireWakeLock()
                startForeground()
                startLocationUpdates()
                Log.i(TAG, "WFH tracking started for EmpId=$empId, Server=$serverUrl")
            }

            ACTION_STOP -> {
                Log.i(TAG, "Stopping WFH tracking for EmpId=$empId")
                stopTracking()
            }
        }

        // START_STICKY: If OS kills the service (memory pressure), restart it automatically
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        stopTracking()
        Log.i(TAG, "WFH Location Service destroyed")
    }

    override fun onBind(intent: Intent?): IBinder? = null

    // ── Foreground Notification ────────────────────────────────────────────────

    private fun startForeground() {
        val notification = buildNotification(
            "WFH GPS tracking is active",
            "Your attendance location is being monitored every 20 seconds"
        )

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }
    }

    private fun buildNotification(title: String, text: String): Notification {
        val tapIntent = Intent(this, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val pendingTap = PendingIntent.getActivity(
            this, 0, tapIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        return NotificationCompat.Builder(this, SmartHRMSApp.WFH_CHANNEL_ID)
            .setContentTitle(title)
            .setContentText(text)
            .setSmallIcon(R.drawable.ic_tracking)
            .setContentIntent(pendingTap)
            .setOngoing(true)           // Cannot be dismissed by swipe
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            // No "Stop Tracking" action — tracking can only be stopped from inside the app
            .setSubText("Ping #$pingCount | ${if (lastPingSuccessful) "OK" else "Connecting..."}")
            .build()
    }

    private fun updateNotification() {
        val manager = getSystemService(NotificationManager::class.java)
        manager.notify(NOTIFICATION_ID, buildNotification(
            "SP OneHR — WFH Tracking Active",
            "Ping #$pingCount | Last: ${java.text.SimpleDateFormat("hh:mm:ss a", java.util.Locale.getDefault()).format(java.util.Date())}"
        ))
    }

    // ── GPS Location Updates ───────────────────────────────────────────────────

    private fun startLocationUpdates() {
        val request = LocationRequest.Builder(
            Priority.PRIORITY_HIGH_ACCURACY,
            20_000L    // Request location every 20 seconds
        ).apply {
            setMinUpdateIntervalMillis(15_000L)    // Fastest: 15 seconds
            setMaxUpdateDelayMillis(30_000L)       // Allow batching up to 30 sec
            setGranularity(Granularity.GRANULARITY_FINE)
            setWaitForAccurateLocation(false)       // Don't wait — use best available
        }.build()

        locationCallback = object : LocationCallback() {
            override fun onLocationResult(result: LocationResult) {
                val location = result.lastLocation ?: return
                lastLat = location.latitude
                lastLon = location.longitude

                Log.d(TAG, "Location update: lat=$lastLat, lon=$lastLon, accuracy=${location.accuracy}m")

                serviceScope.launch {
                    sendPingToServer(
                        lat      = location.latitude,
                        lon      = location.longitude,
                        accuracy = location.accuracy
                    )
                }
            }

            override fun onLocationAvailability(availability: LocationAvailability) {
                if (!availability.isLocationAvailable) {
                    Log.w(TAG, "Location not available — GPS may be disabled")
                }
            }
        }

        try {
            fusedLocationClient.requestLocationUpdates(
                request,
                locationCallback,
                Looper.getMainLooper()
            )
            locationCallbackRegistered = true
            Log.i(TAG, "Location updates registered (every 20 seconds)")
        } catch (e: SecurityException) {
            Log.e(TAG, "Location permission denied: ${e.message}")
            stopSelf()
        }
    }

    private fun stopLocationUpdates() {
        if (locationCallbackRegistered && ::locationCallback.isInitialized) {
            fusedLocationClient.removeLocationUpdates(locationCallback)
            locationCallbackRegistered = false
            Log.i(TAG, "Location updates removed")
        }
    }

    // ── HTTP Ping to Server ────────────────────────────────────────────────────

    private suspend fun sendPingToServer(lat: Double, lon: Double, accuracy: Float) {
        try {
            val deviceModel = "${Build.MANUFACTURER} ${Build.MODEL}"

            val json = JSONObject().apply {
                put("employeeId", empId)
                put("latitude",   lat)
                put("longitude",  lon)
                put("deviceName", deviceModel)
                put("browser",    "SP OneHR Android App v${BuildConfig.VERSION_NAME}")
                put("ipAddress",  "")
            }

            val body = json.toString().toRequestBody("application/json; charset=utf-8".toMediaType())

            val request = Request.Builder()
                .url("$serverUrl/api/wfh/locationPing")
                .post(body)
                .addHeader("User-Agent", "SPOneHR-Android/${BuildConfig.VERSION_NAME}")
                .build()

            httpClient.newCall(request).execute().use { response ->
                pingCount++
                lastPingSuccessful = response.isSuccessful

                if (response.isSuccessful) {
                    val responseBody = response.body?.string() ?: ""
                    Log.d(TAG, "Ping #$pingCount OK | ${response.code} | lat=$lat, lon=$lon | acc=${accuracy}m | $responseBody")
                } else {
                    Log.w(TAG, "Ping #$pingCount failed: HTTP ${response.code}")
                }

                // Update notification with ping count
                withContext(Dispatchers.Main) { updateNotification() }
            }
        } catch (e: Exception) {
            lastPingSuccessful = false
            Log.e(TAG, "Ping error: ${e.javaClass.simpleName} — ${e.message}")
            // Don't crash — just log and continue. Next ping will retry automatically.
        }
    }

    // ── Wake Lock ──────────────────────────────────────────────────────────────

    private fun acquireWakeLock() {
        val pm = getSystemService(POWER_SERVICE) as PowerManager
        wakeLock = pm.newWakeLock(
            PowerManager.PARTIAL_WAKE_LOCK,
            "SPOneHR:WFHTracking"
        ).apply {
            // Acquire indefinitely — released when stopTracking() is called
            acquire(12 * 60 * 60 * 1000L) // Max 12 hours (full workday)
        }
        Log.i(TAG, "Wake lock acquired")
    }

    private fun releaseWakeLock() {
        wakeLock?.let {
            if (it.isHeld) {
                it.release()
                Log.i(TAG, "Wake lock released")
            }
        }
        wakeLock = null
    }

    // ── Stop Everything ────────────────────────────────────────────────────────

    private fun stopTracking() {
        stopLocationUpdates()
        releaseWakeLock()
        serviceScope.cancel()
        stopForeground(STOP_FOREGROUND_REMOVE)
        stopSelf()
    }
}

File: MauiApp/Platforms/Android/WFHLocationService.cs
using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.OS;
using AndroidX.Core.App;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using ALoc = Android.Locations;   // alias avoids ambiguity with MAUI Location type

namespace SPOneHRApp;

[Service(ForegroundServiceType = ForegroundService.TypeLocation)]
public class WFHLocationService : Service, ALoc.ILocationListener
{
    public const string ActionStart    = "com.splotuss.START_WFH_TRACKING";
    public const string ActionStop     = "com.splotuss.STOP_WFH_TRACKING";
    public const string ExtraEmpId     = "emp_id";
    public const string ExtraServerUrl = "server_url";

    private ALoc.LocationManager?    _locationManager;
    private CancellationTokenSource? _cts;

    private readonly HttpClient _httpClient = new HttpClient
    {
        Timeout = TimeSpan.FromSeconds(10)
    };

    private int    _empId       = 0;
    private string _serverUrl   = "";
    private double _lastLat     = 0;
    private double _lastLon     = 0;
    private int    _pingCount   = 0;
    private bool   _hasLocation = false;

    // ── Service Lifecycle ─────────────────────────────────────────────────

    public override void OnCreate()
    {
        base.OnCreate();
        _locationManager = GetSystemService(LocationService) as ALoc.LocationManager;
    }

    public override StartCommandResult OnStartCommand(Intent? intent, StartCommandFlags flags, int startId)
    {
        switch (intent?.Action)
        {
            case ActionStart:
                _empId     = intent.GetIntExtra(ExtraEmpId, 0);
                _serverUrl = intent.GetStringExtra(ExtraServerUrl) ?? AppConstants.ServerUrl;

                if (_empId <= 0) { StopSelf(); return StartCommandResult.NotSticky; }

                StartForeground(AppConstants.NotifId,
                    BuildNotification("WFH Tracking Active", "Initialising GPS..."));
                StartLocationUpdates();
                StartPingLoop();

                Android.Util.Log.Info("WFHService", $"Tracking started — EmpId={_empId}");
                break;

            case ActionStop:
                StopTracking();
                break;
        }

        return StartCommandResult.Sticky;
    }

    public override void OnDestroy()
    {
        StopTracking();
        base.OnDestroy();
    }

    public override IBinder? OnBind(Intent? intent) => null;

    // ── GPS Updates ───────────────────────────────────────────────────────

    private void StartLocationUpdates()
    {
        try
        {
            if (_locationManager?.IsProviderEnabled(ALoc.LocationManager.GpsProvider) == true)
            {
                _locationManager.RequestLocationUpdates(
                    ALoc.LocationManager.GpsProvider,
                    15_000L, 10f, this, Android.OS.Looper.MainLooper);
            }

            if (_locationManager?.IsProviderEnabled(ALoc.LocationManager.NetworkProvider) == true)
            {
                _locationManager.RequestLocationUpdates(
                    ALoc.LocationManager.NetworkProvider,
                    15_000L, 10f, this, Android.OS.Looper.MainLooper);
            }

            Android.Util.Log.Info("WFHService", "Location updates registered");
        }
        catch (Exception ex)
        {
            Android.Util.Log.Error("WFHService", $"Location update error: {ex.Message}");
        }
    }

    // ── ILocationListener ─────────────────────────────────────────────────

    public void OnLocationChanged(ALoc.Location location)
    {
        _lastLat     = location.Latitude;
        _lastLon     = location.Longitude;
        _hasLocation = true;
        Android.Util.Log.Debug("WFHService",
            $"GPS: {_lastLat:F6}, {_lastLon:F6} acc={location.Accuracy:F0}m");
    }

    public void OnProviderEnabled(string provider)  =>
        Android.Util.Log.Info("WFHService", $"Provider enabled: {provider}");

    public void OnProviderDisabled(string provider) =>
        Android.Util.Log.Warn("WFHService", $"Provider disabled: {provider}");

    // Required by ILocationListener — no-op for modern Android
    public void OnStatusChanged(string? provider, ALoc.Availability status, Android.OS.Bundle? extras) { }

    // ── Ping Loop ─────────────────────────────────────────────────────────

    private void StartPingLoop()
    {
        _cts = new CancellationTokenSource();
        _ = PingLoop(_cts.Token);
    }

    private async Task PingLoop(CancellationToken ct)
    {
        await Task.Delay(3_000, ct);  // warm-up for first GPS fix

        while (!ct.IsCancellationRequested)
        {
            if (_hasLocation)
                await SendPing(_lastLat, _lastLon, ct);

            UpdateNotification();
            await Task.Delay(AppConstants.PingIntervalMs, ct);
        }
    }

    private async Task SendPing(double lat, double lon, CancellationToken ct)
    {
        try
        {
            var payload = new
            {
                employeeId = _empId,
                latitude   = lat,
                longitude  = lon,
                deviceName = $"{Android.OS.Build.Manufacturer} {Android.OS.Build.Model}",
                browser    = $"SP OneHR MAUI v{AppConstants.AppVersion}",
                ipAddress  = ""
            };

            var json = JsonSerializer.Serialize(payload);
            using var content  = new StringContent(json, Encoding.UTF8, "application/json");
            using var response = await _httpClient.PostAsync(
                $"{_serverUrl}/api/wfh/locationPing", content, ct);

            _pingCount++;
            Android.Util.Log.Debug("WFHService",
                $"Ping #{_pingCount} HTTP {(int)response.StatusCode} | {lat:F5},{lon:F5}");
        }
        catch (Exception ex) when (!ct.IsCancellationRequested)
        {
            Android.Util.Log.Warn("WFHService", $"Ping error: {ex.Message}");
        }
    }

    // ── Notification ──────────────────────────────────────────────────────

    private Notification BuildNotification(string title, string text)
    {
        CreateNotificationChannel();

        var tapIntent = new Intent(this, typeof(MainActivity));
        var pendingTap = PendingIntent.GetActivity(this, 0, tapIntent,
            PendingIntentFlags.UpdateCurrent | PendingIntentFlags.Immutable)!;

        var stopIntent = new Intent(this, typeof(WFHLocationService));
        stopIntent.SetAction(ActionStop);
        var pendingStop = PendingIntent.GetService(this, 1, stopIntent,
            PendingIntentFlags.UpdateCurrent | PendingIntentFlags.Immutable)!;

        return new NotificationCompat.Builder(this, AppConstants.NotifChannelId)
            .SetContentTitle(title)
            .SetContentText(text)
            .SetSmallIcon(Resource.Drawable.ic_notification)
            .SetContentIntent(pendingTap)
            .AddAction(Resource.Drawable.ic_notification, "Stop", pendingStop)
            .SetOngoing(true)
            .SetPriority(NotificationCompat.PriorityLow)
            .SetVisibility(NotificationCompat.VisibilityPublic)
            .Build();
    }

    private void UpdateNotification()
    {
        var mgr = GetSystemService(NotificationService) as Android.App.NotificationManager;
        mgr?.Notify(AppConstants.NotifId, BuildNotification(
            "SP OneHR — WFH Tracking",
            _hasLocation
                ? $"Ping #{_pingCount} | {_lastLat:F4}, {_lastLon:F4}"
                : "Acquiring GPS..."));
    }

    private void CreateNotificationChannel()
    {
        if (Android.OS.Build.VERSION.SdkInt >= Android.OS.BuildVersionCodes.O)
        {
            var channel = new Android.App.NotificationChannel(
                AppConstants.NotifChannelId,
                AppConstants.NotifChannelName,
                Android.App.NotificationImportance.Low)
            {
                Description = "Shows while WFH GPS tracking is active"
            };
            channel.SetShowBadge(false);
            var mgr = GetSystemService(NotificationService) as Android.App.NotificationManager;
            mgr?.CreateNotificationChannel(channel);
        }
    }

    // ── Stop ──────────────────────────────────────────────────────────────

    private void StopTracking()
    {
        _cts?.Cancel();
        _locationManager?.RemoveUpdates(this);
        StopForeground(StopForegroundFlags.Remove);
        StopSelf();
        Android.Util.Log.Info("WFHService", "Tracking stopped");
    }
}
